File: Python/juegoconherencia/Mago.py
from Personaje import Personaje
import random

class Mago(Personaje):
    
    VIDA_POR_DEFECTO = 100
    
    def __init__(self, nombre: str, vida: int):
        super().__init__(nombre, vida)
    
    @classmethod    
    def Constructor(cls, nombre):
        nuevoMago = cls(nombre, cls.VIDA_POR_DEFECTO)
        return nuevoMago
    
    def recibir_dano(self, dano: int):
        if random.randint(0,2) == 1:
            dano = self.__habilidad_especial__(dano)
        super().recibir_dano(dano)
        
    def __habilidad_especial__(self, dano):
        return int(dano*0.2)

File: Java/juegoconherencia/src/Mago.java

import java.util.Random;


public class Mago extends Personaje{
    
    private static final int VIDA_POR_DEFECTO = 100;

    public Mago(String nombre, int vida) {
        super(nombre, vida);
    }

    public Mago(String nombre) {
        this(nombre, VIDA_POR_DEFECTO);
    }

    @Override
    public void recibirDano(int dano) {
        if(new Random().nextBoolean()){
            dano  = this.hablidadEspecial(dano);
        }
        super.recibirDano(dano);
    }

    @Override
    protected Integer hablidadEspecial(int dano) {
        return (int) (dano*0.2);
    }
    
}
